File: scripts/nats_extract_examples.py
#!/usr/bin/env python3
"""
GitHub webhook example extractor for the HephaestusTest organization.
Extracts webhook payload examples from NATS stream and saves them as JSON files.
"""

import argparse
import asyncio
import json
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, Optional, Set
import nats
from nats.errors import TimeoutError
from nats.js import api as jsapi

# Default configuration
SCRIPT_DIR = Path(__file__).resolve().parent
REPO_ROOT = SCRIPT_DIR.parent
DEFAULT_NATS_SERVER = "nats://131.159.89.80:4222"
DEFAULT_EXAMPLES_DIR = REPO_ROOT / "server" / "application-server" / "src" / "test" / "resources" / "github"
DEFAULT_NATS_SUBJECT = "github.HephaestusTest.>"
DEFAULT_NATS_STREAM = "github"


def get_example_filename(event_type: str, action: str = None) -> str:
    """
    Generate filename for webhook example based on event type and action.
    Format: <event_type>.<action>.json or <event_type>.json (if no action)
    """
    if action:
        return f"{event_type}.{action}.json"
    else:
        return f"{event_type}.json"


def get_existing_examples(examples_dir: Path) -> Set[str]:
    """Get set of existing example filenames"""
    existing = set()
    if examples_dir.exists():
        for file in examples_dir.glob("*.json"):
            existing.add(file.name)
    return existing


async def extract_webhook_examples(
    nats_server: str,
    examples_dir: Path,
    nats_subject: str,
    nats_stream: str,
    event_filters: Dict[str, Set[str]],
    since: Optional[datetime],
    until: Optional[datetime],
    allow_duplicates: bool,
    start_with_new: bool,
    batch_size: int,
    fetch_timeout: float
):
    """
    Main extraction function.
    Consumes all available messages from HephaestusTest and extracts unique examples.
    """
    print("🚀 Starting webhook example extraction...")
    print(f"📡 Connecting to NATS server: {nats_server}")
    print(f"📁 Output directory: {examples_dir}")
    print(f"🎯 Subject pattern: {nats_subject}")
    
    # Ensure output directory exists
    examples_dir.mkdir(parents=True, exist_ok=True)
    
    # Track existing examples to avoid duplicates
    existing_examples = get_existing_examples(examples_dir)
    extracted_examples = {}  # filename -> payload
    filename_counts: Dict[str, int] = {}
    processed_count = 0
    matched_count = 0
    skipped_by_filter = 0
    skipped_by_time = 0
    
    try:
        # Connect to NATS
        nc = await nats.connect(nats_server)
        js = nc.jetstream()
    except Exception as e:
        print(f"❌ Failed to connect to NATS server: {e}")
        return False

    psub = None
    try:
        print(f"📊 Found {len(existing_examples)} existing examples")
        
        # First, let's check what messages are available in the stream
        try:
            stream_info = await js.stream_info(nats_stream)
            print(f"📊 Stream info: {stream_info.state.messages} total messages")
        except Exception as e:
            print(f"⚠️  Could not get stream info: {e}")

        # Create a pull subscription to consume all available messages
        # Use ephemeral consumer to start from beginning every time
        consumer_config = jsapi.ConsumerConfig(
            ack_policy=jsapi.AckPolicy.EXPLICIT,
            deliver_policy=jsapi.DeliverPolicy.ALL,
        )
        if start_with_new:
            consumer_config.deliver_policy = jsapi.DeliverPolicy.NEW
            print("⚙️  Consumer deliver policy: NEW (future messages only)")
        elif since:
            consumer_config.deliver_policy = jsapi.DeliverPolicy.BY_START_TIME
            start_time = since.astimezone(timezone.utc)
            consumer_config.opt_start_time = start_time.isoformat()
            print(f"⚙️  Consumer deliver policy: BY_START_TIME from {start_time.isoformat()}")
        else:
            print("⚙️  Consumer deliver policy: ALL (full stream history)")

        try:
            psub = await js.pull_subscribe(
                subject=nats_subject,
                stream=nats_stream,
                config=consumer_config,
            )
            print(f"✅ Successfully created subscription for: {nats_subject}")
        except Exception as e:
            print(f"❌ Failed to create subscription: {e}")
            return False
        
        print("🔄 Consuming messages...")
        
        # Consume all available messages
        while True:
            try:
                # Fetch messages in batches
                msgs = await psub.fetch(batch_size, timeout=fetch_timeout)
                
                if not msgs:
                    print("ℹ️  No more messages available")
                    break
                
                for msg in msgs:
                    try:
                        # Parse the webhook payload
                        payload = json.loads(msg.data.decode())
                        processed_count += 1
                        
                        # Extract event type from NATS subject
                        # Subject format: github.{org}.{repo}.{event_type}
                        subject_parts = msg.subject.split('.')
                        if len(subject_parts) >= 4:
                            event_type = subject_parts[3]
                            normalized_event_type = event_type.lower()
                        else:
                            print(f"   ⚠️  Unexpected subject format: {msg.subject}")
                            await msg.ack()
                            continue
                        
                        # Extract action from payload if present
                        action = payload.get("action")

                        # Skip events that do not match requested filters
                        if event_filters:
                            allowed_actions = event_filters.get(normalized_event_type)
                            if allowed_actions is None:
                                skipped_by_filter += 1
                                await msg.ack()
                                continue
                            normalized_action = (action or "").lower()
                            if allowed_actions and normalized_action not in allowed_actions:
                                skipped_by_filter += 1
                                await msg.ack()
                                continue
                            matched_count += 1
                        
                        metadata = getattr(msg, "metadata", None)
                        msg_timestamp = getattr(metadata, "timestamp", None)

                        if since and msg_timestamp and msg_timestamp < since:
                            skipped_by_time += 1
                            await msg.ack()
                            continue
                        if until and msg_timestamp and msg_timestamp > until:
                            skipped_by_time += 1
                            await msg.ack()
                            continue

                        # Generate filename for this webhook type
                        filename = get_example_filename(event_type, action)

                        if allow_duplicates:
                            base_name = filename[:-5]  # drop .json
                            count = filename_counts.get(base_name, 0)
                            candidate = filename

                            while (
                                candidate in existing_examples
                                or candidate in extracted_examples
                            ):
                                count += 1
                                candidate = f"{base_name}.{count}.json"

                            filename = candidate
                            filename_counts[base_name] = count
                        else:
                            if filename in existing_examples or filename in extracted_examples:
                                await msg.ack()
                                continue

                        # Store the example
                        extracted_examples[filename] = payload
                        print(f"   ✅ Found new example: {filename}")
                        
                        await msg.ack()
                        
                    except json.JSONDecodeError:
                        print(f"   ⚠️  Skipping invalid JSON message")
                        await msg.ack()
                        continue
                    except Exception as e:
                        print(f"   ⚠️  Error processing message: {e}")
                        await msg.ack()
                        continue
                
            except TimeoutError:
                print("ℹ️  No more messages available (timeout)")
                break
            except Exception as e:
                print(f"⚠️  Error fetching messages: {e}")
                break
        
        # Write all extracted examples to files
        for filename, payload in extracted_examples.items():
            filepath = examples_dir / filename
            try:
                with open(filepath, 'w') as f:
                    json.dump(payload, f, indent=2)
            except Exception as e:
                print(f"❌ Failed to write {filename}: {e}")
        
        print(f"\n🎉 Extraction complete!")
        print(f"📊 Processed: {processed_count} messages")
        if event_filters:
            print(f"📊 Matched filter: {matched_count} messages")
            if skipped_by_filter:
                print(f"📊 Skipped by event filter: {skipped_by_filter} messages")
        if skipped_by_time:
            print(f"📊 Skipped by time window: {skipped_by_time} messages")
        print(f"📊 Extracted: {len(extracted_examples)} new examples")
        print(f"📁 Total examples: {len(existing_examples) + len(extracted_examples)}")
        
        if extracted_examples:
            print(f"\n📝 New examples created:")
            for filename in sorted(extracted_examples.keys()):
                print(f"   - {filename}")
        
        return True
        
    finally:
        try:
            if psub is not None:
                await psub.unsubscribe()
        except Exception:
            pass
        await nc.close()


def parse_event_filters(raw_filters) -> Dict[str, Set[str]]:
    """Parse --event arguments into an event -> actions mapping."""
    filters: Dict[str, Set[str]] = {}
    for raw in raw_filters or []:
        parts = raw.split(":", 1)
        event = parts[0].strip().lower()
        if not event:
            continue
        action_set = filters.setdefault(event, set())
        if len(parts) == 2:
            action = parts[1].strip().lower()
            if action:
                action_set.add(action)
        else:
            # Empty set indicates all actions for this event
            filters[event] = action_set
    return filters

def parse_iso_datetime(value: Optional[str]) -> Optional[datetime]:
    """Parse ISO8601 string into timezone-aware datetime."""
    if not value:
        return None
    cleaned = value.strip()
    if not cleaned:
        return None
    if cleaned.endswith("Z"):
        cleaned = cleaned[:-1] + "+00:00"
    dt = datetime.fromisoformat(cleaned)
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt


def main():
    """Main entry point with argument parsing"""
    parser = argparse.ArgumentParser(
        description="Extract GitHub webhook examples from NATS stream"
    )
    parser.add_argument(
        "--nats-server",
        default=DEFAULT_NATS_SERVER,
        help=f"NATS server URL (default: {DEFAULT_NATS_SERVER})"
    )
    parser.add_argument(
        "--examples-dir",
        type=Path,
        default=DEFAULT_EXAMPLES_DIR,
        help=f"Output directory for examples (default: {DEFAULT_EXAMPLES_DIR})"
    )
    parser.add_argument(
        "--subject",
        default=DEFAULT_NATS_SUBJECT,
        help=f"NATS subject pattern (default: {DEFAULT_NATS_SUBJECT})"
    )
    parser.add_argument(
        "--stream",
        default=DEFAULT_NATS_STREAM,
        help=f"NATS stream name (default: {DEFAULT_NATS_STREAM})"
    )
    parser.add_argument(
        "--event",
        action="append",
        dest="events",
        help=(
            "Filter to specific event/action pairs. "
            "Format: event[:action]. Use multiple --event flags for more pairs."
        )
    )
    parser.add_argument(
        "--since",
        help=(
            "Only include messages published at or after this ISO8601 timestamp "
            "(e.g. 2025-10-28T22:00:00+01:00)"
        )
    )
    parser.add_argument(
        "--until",
        help=(
            "Only include messages published at or before this ISO8601 timestamp"
        )
    )
    parser.add_argument(
        "--allow-duplicates",
        action="store_true",
        help=(
            "Allow multiple payloads per event/action and append numeric suffixes "
            "to filenames when necessary"
        )
    )
    parser.add_argument(
        "--start-with-new",
        action="store_true",
        help=(
            "Subscribe only to new messages. Useful when you do not need the full "
            "stream history and want to avoid scanning millions of older events."
        )
    )
    parser.add_argument(
        "--batch-size",
        type=int,
        default=50,
        help="Number of messages to request per fetch call (default: 50)"
    )
    parser.add_argument(
        "--fetch-timeout",
        type=float,
        default=5.0,
        help="Seconds to wait for a fetch batch before timing out (default: 5.0)"
    )
    
    args = parser.parse_args()
    event_filters = parse_event_filters(args.events)
    since = parse_iso_datetime(args.since)
    until = parse_iso_datetime(args.until)
    if args.batch_size <= 0:
        parser.error("--batch-size must be a positive integer")
    if args.fetch_timeout <= 0:
        parser.error("--fetch-timeout must be positive")

    examples_dir = args.examples_dir.expanduser().resolve()
    
    try:
        success = asyncio.run(extract_webhook_examples(
            args.nats_server,
            examples_dir,
            args.subject,
            args.stream,
            event_filters,
            since,
            until,
            args.allow_duplicates,
            args.start_with_new,
            args.batch_size,
            args.fetch_timeout
        ))
        
        if not success:
            sys.exit(1)
            
    except KeyboardInterrupt:
        print("\n👋 Extraction cancelled by user")
        sys.exit(0)
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/nats-extract-examples.ts
#!/usr/bin/env node
import fs from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";
import type {
	ConsumerConfig,
	ConsumerMessages,
	JsMsg,
} from "@nats-io/jetstream";
import {
	AckPolicy,
	DeliverPolicy,
	jetstream,
	jetstreamManager,
	ReplayPolicy,
} from "@nats-io/jetstream";
import type { NatsConnection } from "@nats-io/transport-node";
import { connect } from "@nats-io/transport-node";
import { Command, InvalidArgumentError, Option } from "commander";

type LogLevel = "debug" | "info" | "silent";

type Logger = {
	debug: (message: string) => void;
	info: (message: string) => void;
	error: (message: string) => void;
};

function createLogger(level: LogLevel): Logger {
	return {
		debug: (msg) => level === "debug" && console.log(msg),
		info: (msg) => level !== "silent" && console.log(msg),
		error: (msg) => console.error(msg),
	};
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const REPO_ROOT = path.resolve(__dirname, "..");

const DEFAULT_NATS_SERVER = process.env.NATS_URL ?? "nats://localhost:4222";
const DEFAULT_EXAMPLES_DIR = path.join(
	REPO_ROOT,
	"server",
	"application-server",
	"src",
	"test",
	"resources",
	"github",
);
const DEFAULT_NATS_SUBJECT = "github.HephaestusTest.>";
const DEFAULT_NATS_STREAM = "github";

function parsePositiveInt(value: string): number {
	const parsed = Number.parseInt(value, 10);
	if (!Number.isFinite(parsed) || parsed <= 0) {
		throw new InvalidArgumentError(`Expected a positive integer, got ${value}`);
	}
	return parsed;
}

function parsePositiveFloat(value: string): number {
	const parsed = Number.parseFloat(value);
	if (!Number.isFinite(parsed) || parsed <= 0) {
		throw new InvalidArgumentError(`Expected a positive number, got ${value}`);
	}
	return parsed;
}

function parseLogLevel(value: string): LogLevel {
	if (value === "debug" || value === "info" || value === "silent") {
		return value;
	}
	throw new InvalidArgumentError(`Invalid log level: ${value}`);
}

function parseIsoDate(value: string): Date {
	const trimmed = value.trim();
	if (!trimmed) {
		throw new InvalidArgumentError("Timestamp cannot be empty");
	}
	const hasOffset = /[+-]\d{2}:\d{2}$/.test(trimmed);
	const normalized =
		trimmed.endsWith("Z") || hasOffset ? trimmed : `${trimmed}Z`;
	const parsed = new Date(normalized);
	if (Number.isNaN(parsed.getTime())) {
		throw new InvalidArgumentError(`Invalid ISO8601 timestamp: ${value}`);
	}
	return parsed;
}

function parseEventFilters(entries: string[]): Map<string, Set<string>> {
	const filters = new Map<string, Set<string>>();
	for (const entry of entries) {
		const [eventRaw, actionRaw] = entry.split(":", 2);
		const event = eventRaw?.trim().toLowerCase();
		if (!event) {
			continue;
		}
		const existing = filters.get(event) ?? new Set<string>();
		if (actionRaw) {
			const action = actionRaw.trim().toLowerCase();
			if (action) {
				existing.add(action);
			}
		}
		filters.set(event, existing);
	}
	return filters;
}

function getExampleFilename(
	eventType: string,
	action: string | undefined,
): string {
	return action ? `${eventType}.${action}.json` : `${eventType}.json`;
}

async function getExistingExamples(examplesDir: string): Promise<Set<string>> {
	try {
		const entries = await fs.readdir(examplesDir, { withFileTypes: true });
		return new Set(
			entries
				.filter((entry) => entry.isFile() && entry.name.endsWith(".json"))
				.map((entry) => entry.name),
		);
	} catch {
		return new Set();
	}
}

function getMsgTimestamp(msg: JsMsg): Date | null {
	const timestamp = msg.time;
	return Number.isNaN(timestamp.getTime()) ? null : timestamp;
}

type ExtractOptions = {
	natsServer: string;
	examplesDir: string;
	natsSubject: string;
	natsStream: string;
	eventFilters: Map<string, Set<string>>;
	since: Date | null;
	until: Date | null;
	allowDuplicates: boolean;
	startWithNew: boolean;
	batchSize: number;
	fetchTimeoutMs: number;
	dryRun: boolean;
	logLevel: LogLevel;
};

async function extractWebhookExamples(options: ExtractOptions, logger: Logger) {
	logger.info(`Connecting to NATS server: ${options.natsServer}`);
	logger.info(`Output directory: ${options.examplesDir}`);
	logger.info(`Subject pattern: ${options.natsSubject}`);

	await fs.mkdir(options.examplesDir, { recursive: true });

	const existingExamples = await getExistingExamples(options.examplesDir);
	const extractedExamples = new Map<string, unknown>();
	const filenameCounts = new Map<string, number>();

	let processedCount = 0;
	let matchedCount = 0;
	let skippedByFilter = 0;
	let skippedByTime = 0;

	const decoder = new TextDecoder();
	let nc: NatsConnection | null = null;
	let consumerName: string | null = null;
	let consumerRef: { delete: () => Promise<boolean> } | null = null;
	try {
		nc = await connect({ servers: options.natsServer });
		const js = jetstream(nc);
		const jsm = await jetstreamManager(nc);

		try {
			const streamInfo = await jsm.streams.info(options.natsStream);
			logger.info(
				`Stream info: ${streamInfo.state.messages} total messages in ${options.natsStream}`,
			);
		} catch (error) {
			logger.info(`Could not get stream info: ${(error as Error).message}`);
		}

		const deliverPolicy = options.startWithNew
			? DeliverPolicy.New
			: options.since
				? DeliverPolicy.StartTime
				: DeliverPolicy.All;

		if (deliverPolicy === DeliverPolicy.New) {
			logger.info("Consumer deliver policy: NEW (future messages only)");
		} else if (deliverPolicy === DeliverPolicy.StartTime && options.since) {
			logger.info(
				`Consumer deliver policy: START_TIME from ${options.since.toISOString()}`,
			);
		} else {
			logger.info("Consumer deliver policy: ALL (full stream history)");
		}

		consumerName = `extract-examples-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
		const consumerConfig: ConsumerConfig = {
			name: consumerName,
			ack_policy: AckPolicy.Explicit,
			deliver_policy: deliverPolicy,
			replay_policy: ReplayPolicy.Instant,
			filter_subject: options.natsSubject,
		};
		if (deliverPolicy === DeliverPolicy.StartTime && options.since) {
			consumerConfig.opt_start_time = options.since.toISOString();
		}

		await jsm.consumers.add(options.natsStream, consumerConfig);
		const consumer = await js.consumers.get(options.natsStream, consumerName);
		consumerRef = consumer;

		logger.info(`Consumer created: ${consumerName}`);

		while (true) {
			let msgs: ConsumerMessages;
			try {
				msgs = await consumer.fetch({
					max_messages: options.batchSize,
					expires: options.fetchTimeoutMs,
				});
			} catch (error) {
				const err = error as Error & { code?: string; name?: string };
				if (err.name === "TimeoutError" || err.code === "TIMEOUT") {
					logger.info("No more messages available (timeout)");
					break;
				}
				throw error;
			}

			let gotAny = false;
			try {
				for await (const msg of msgs) {
					gotAny = true;
					processedCount += 1;

					let payload: Record<string, unknown>;
					try {
						payload = JSON.parse(decoder.decode(msg.data)) as Record<
							string,
							unknown
						>;
					} catch {
						logger.info("Skipping invalid JSON message");
						msg.ack();
						continue;
					}

					const subjectParts = msg.subject.split(".");
					if (subjectParts.length < 4) {
						logger.info(`Unexpected subject format: ${msg.subject}`);
						msg.ack();
						continue;
					}

					const eventType = subjectParts[3];
					if (!eventType) {
						logger.info(`Unexpected subject format: ${msg.subject}`);
						msg.ack();
						continue;
					}
					const normalizedEventType = eventType.toLowerCase();
					const action = payload.action ? String(payload.action) : undefined;

					if (options.eventFilters.size > 0) {
						const allowedActions =
							options.eventFilters.get(normalizedEventType);
						if (!allowedActions) {
							skippedByFilter += 1;
							msg.ack();
							continue;
						}
						if (allowedActions.size > 0) {
							const normalizedAction = (action ?? "").toLowerCase();
							if (!allowedActions.has(normalizedAction)) {
								skippedByFilter += 1;
								msg.ack();
								continue;
							}
						}
						matchedCount += 1;
					}

					const msgTimestamp = getMsgTimestamp(msg);
					if (options.since && msgTimestamp && msgTimestamp < options.since) {
						skippedByTime += 1;
						msg.ack();
						continue;
					}
					if (options.until && msgTimestamp && msgTimestamp > options.until) {
						skippedByTime += 1;
						msg.ack();
						continue;
					}

					let filename = getExampleFilename(eventType, action);
					if (options.allowDuplicates) {
						const baseName = filename.replace(/\.json$/, "");
						const count = filenameCounts.get(baseName) ?? 0;
						let candidate = filename;
						let counter = count;
						while (
							existingExamples.has(candidate) ||
							extractedExamples.has(candidate)
						) {
							counter += 1;
							candidate = `${baseName}.${counter}.json`;
						}
						filenameCounts.set(baseName, counter);
						filename = candidate;
					} else if (
						existingExamples.has(filename) ||
						extractedExamples.has(filename)
					) {
						msg.ack();
						continue;
					}

					extractedExamples.set(filename, payload);
					logger.debug(`Found new example: ${filename}`);
					msg.ack();
				}
			} catch (error) {
				logger.error(`Error fetching messages: ${(error as Error).message}`);
				break;
			}

			if (!gotAny) {
				logger.info("No more messages available");
				break;
			}
		}

		for (const [filename, payload] of extractedExamples.entries()) {
			const filepath = path.join(options.examplesDir, filename);
			await fs.writeFile(filepath, JSON.stringify(payload, null, 2));
		}

		logger.info("Extraction complete.");
		logger.info(`Processed: ${processedCount} messages`);
		if (options.eventFilters.size > 0) {
			logger.info(`Matched filter: ${matchedCount} messages`);
			if (skippedByFilter > 0) {
				logger.info(`Skipped by event filter: ${skippedByFilter} messages`);
			}
		}
		if (skippedByTime > 0) {
			logger.info(`Skipped by time window: ${skippedByTime} messages`);
		}
		logger.info(`Extracted: ${extractedExamples.size} new examples`);
		logger.info(
			`Total examples: ${existingExamples.size + extractedExamples.size}`,
		);

		if (extractedExamples.size > 0) {
			logger.info("New examples created:");
			for (const filename of [...extractedExamples.keys()].sort()) {
				logger.info(` - ${filename}`);
			}
		}
	} finally {
		try {
			if (consumerRef) {
				await consumerRef.delete();
				logger.debug(`Consumer deleted: ${consumerName ?? "unknown"}`);
			}
		} catch {
			// ignore
		}
		if (nc) {
			await nc.close();
		}
	}
}

function buildProgram() {
	const program = new Command();
	program
		.name("nats-extract-examples")
		.description("Extract webhook examples from NATS JetStream")
		.option("--nats-server <url>", "NATS server URL", DEFAULT_NATS_SERVER)
		.option("--examples-dir <path>", "Output directory", DEFAULT_EXAMPLES_DIR)
		.option("--subject <subject>", "NATS subject pattern", DEFAULT_NATS_SUBJECT)
		.option("--stream <stream>", "NATS stream name", DEFAULT_NATS_STREAM)
		.option(
			"--event <event[:action]>",
			"Filter by event/action",
			(value, previous) => {
				return [...previous, value];
			},
			[] as string[],
		)
		.option("--since <iso>", "Only include messages after this timestamp")
		.option("--until <iso>", "Only include messages before this timestamp")
		.option(
			"--allow-duplicates",
			"Allow multiple examples per event/action",
			false,
		)
		.option("--start-with-new", "Consume only new messages", false)
		.option("--batch-size <n>", "Batch size per fetch", parsePositiveInt, 50)
		.option(
			"--fetch-timeout <sec>",
			"Fetch timeout in seconds",
			parsePositiveFloat,
			5,
		)
		.option("--dry-run", "Validate configuration and exit", false)
		.addOption(
			new Option("--log-level <level>", "Log verbosity")
				.choices(["debug", "info", "silent"])
				.default("info")
				.argParser(parseLogLevel),
		);
	return program;
}

async function main() {
	const program = buildProgram();
	program.parse(process.argv);
	const rawOptions = program.opts<{
		natsServer: string;
		examplesDir: string;
		subject: string;
		stream: string;
		event: string[];
		since?: string;
		until?: string;
		allowDuplicates: boolean;
		startWithNew: boolean;
		batchSize: number;
		fetchTimeout: number;
		dryRun: boolean;
		logLevel: LogLevel;
	}>();

	const logger = createLogger(rawOptions.logLevel);

	if (rawOptions.startWithNew && (rawOptions.since || rawOptions.until)) {
		throw new InvalidArgumentError(
			"--start-with-new cannot be combined with --since/--until",
		);
	}

	const since = rawOptions.since ? parseIsoDate(rawOptions.since) : null;
	const until = rawOptions.until ? parseIsoDate(rawOptions.until) : null;
	if (since && until && since > until) {
		throw new InvalidArgumentError("--since must be earlier than --until");
	}

	const options: ExtractOptions = {
		natsServer: rawOptions.natsServer,
		examplesDir: path.resolve(rawOptions.examplesDir),
		natsSubject: rawOptions.subject,
		natsStream: rawOptions.stream,
		eventFilters: parseEventFilters(rawOptions.event),
		since,
		until,
		allowDuplicates: rawOptions.allowDuplicates,
		startWithNew: rawOptions.startWithNew,
		batchSize: rawOptions.batchSize,
		fetchTimeoutMs: Math.round(rawOptions.fetchTimeout * 1000),
		dryRun: rawOptions.dryRun,
		logLevel: rawOptions.logLevel,
	};

	logger.debug(
		`Resolved options: ${JSON.stringify({
			natsServer: options.natsServer,
			examplesDir: options.examplesDir,
			natsSubject: options.natsSubject,
			natsStream: options.natsStream,
			eventFilters: [...options.eventFilters.entries()].map(
				([event, actions]) => ({
					event,
					actions: [...actions],
				}),
			),
			since: options.since?.toISOString() ?? null,
			until: options.until?.toISOString() ?? null,
			allowDuplicates: options.allowDuplicates,
			startWithNew: options.startWithNew,
			batchSize: options.batchSize,
			fetchTimeoutMs: options.fetchTimeoutMs,
			dryRun: options.dryRun,
			logLevel: options.logLevel,
		})}`,
	);

	if (options.dryRun) {
		logger.info("Dry run enabled. Configuration validated successfully.");
		return;
	}

	await extractWebhookExamples(options, logger);
}

main().catch((error) => {
	const message = error instanceof Error ? error.message : String(error);
	console.error(`Webhook extraction failed: ${message}`);
	process.exit(1);
});
